Fix a date-time bug; drop pointless dependencies; and deprecate utils.

dmyersturnbull opened the pull request:
@piehld Another PR will remove MySQL. This PR fixes a bug and removes a few unnecessary dependencies.

Fixes a date-time calculation bug; drops pointless date-time dependencies; and deprecates TimeUtil and TextUtil.

1. `TimeUtil` incorrectly called `datetime.replace(tzinfo=utc)`, which does doesn't actually move timezones. Code using `TimeUtil` and `useUtc` was affected.
2. Removes the dependencies [`dateutil`](https://pypi.org/project/python-dateutil/), [`pytz`](https://pypi.org/project/pytz/), and [`strict-rfc3339`](https://pypi.org/project/strict-rfc3339/). Builtin `datetime` works perfectly well. (pytz has some cool stuff, though.) Removes these dependencies from `TimeUtil` and some other code.
3. Refactored code using `TimeUtil` to avoid using it. Its methods just delegate to `datetime` functions in standard ways. [note 1]
4. Cleaned up `TextUtil` and refactored code to avoid using it. It only defined `unescapeXmlCharRef`, which is just `html.unescape` in Python 3.4+.
5. Deleted the file that was an unused duplication of `TextUtil`. [note 2]

[note 1] The only non-obvious code is the "week of the year signature".
It's nonstandard (i.e. not [ISO 8601 week number](https://en.wikipedia.org/wiki/ISO_week_date)). However, it's literally just `strftime("%Y_%V")`.

[note 2] This is technically a breaking change, but it's not being used. If outside code imports it, we could add

```python
from TextUtil import unescapeXmlCharRef as whatever`
```

piehld reviewed (changes requested):
Thank @dmyersturnbull, definitely appreciate you reviewing that codebase and identifying and proposing fixes to outdated code. I'm sure your changes are reliable and will try to review them soon, but because I'm out the rest of the week, I just want to give you a heads up that we will probably wait until next week to merge this, since I want to be extra careful testing this before pushing out to production for the weekend. Any changes with `rcsb.db` are pretty impactful with the weekly-update, so I just tend to lean towards extra caution when it comes to pushing things out is all. 

While I still need to go through each change, the one thing that stands out to me is the removal of the `mock-data` submodule. I'm guessing that wasn't intentional...can you add it back in there?

Also FYI @brindakv

dmyersturnbull commented:
> While I still need to go through each change, the one thing that stands out to me is the removal of the mock-data submodule. I'm guessing that wasn't intentional...can you add it back in there?

Ah! Git doesn't normally track empty directories, and I had to `rmdir` to bypass an error. Apparently that deletes submodules. I'll fix it.

piehld commented:
Thanks! And be aware that Michael just merged a PR, so you'll probably want to rebase or merge.

> By the way, do tests run for PRs?

Yes, they do. Every time you push a new commit a new pipeline will run. However, for some reason this repository doesn't show the results from the runs like other repos do, so you'll need to just go to Azure to check the run directly: https://dev.azure.com/rcsb/RCSB%20PDB%20Python%20Projects/_build?definitionId=33

When you push a new commit, a new run will show up on that link.

dmyersturnbull commented:
@piehld I rebased. I also adjusted the readme (no longer mentions MySQL) and cleaned up `setup.py`. I moved `rcsb.utils.chemref` to the test requirements.

piehld commented on rcsb/db/utils/TimeUtil.py:
Alternatively, instead of getting rid of this file (since it is used in a handful of places), is it possible to just update this file to not use `dateutil` or `pytz`? That would be my preference, at least.

piehld replied:
Hmm, I noticed that difference too, but I also see imports of `rcsb.db.utils.TimeUtil`, e.g.:
- https://github.com/rcsb/py-rcsb_exdb/blob/master/rcsb/exdb/utils/ObjectTransformer.py#L20
- https://github.com/rcsb/py-rcsb_exdb/blob/master/rcsb/exdb/utils/ObjectValidator.py#L24
- https://github.com/rcsb/py-rcsb_workflow/blob/master/rcsb/workflow/wuw/ExDbWorkflow.py#L23

dmyersturnbull replied:
Oh, good catch. I added stubs to `rcsb/db/utils/__init__.py` that provide those with deprecation warnings.

piehld commented on rcsb/db/processors/DataExchangeStatus.py:
```diff
@@ -15,8 +15,9 @@
 import logging
-
-from rcsb.db.utils.TimeUtil import TimeUtil
+from datetime import datetime
+from typing import Optional
+from zoneinfo import ZoneInfo
```
Likewise, if we keep `TimeUtil`, we don't have to worry about updating each of these files that currently imports it.

piehld commented on rcsb/db/processors/DataTransformFactory.py:
```diff
@@ -384,37 +383,54 @@ def castDateToObj(self, trfTup):
+        Returns:
+            A TrfValue tuple
         """
+        # TODO
+        # What is this function doing???!
+        # Its name was castDateTimeToIsoDate, suggesting it outputs YYYY-MM-DD.
+        # Contradicting tht, the docstring said:
+        # > Cast the input date (optional time) string (yyyy-mm-dd:hh::mm:ss) to a Python DateTime object
+        # Instead, it did this:
+        # > tS = dateutil.parser.parse(tv).replace(tzinfo=pytz.UTC).isoformat()
+        # That's three contradictory claims.
+        # What was this function supposed to do?
+        # Why did it fill in 00:00:00, and why did it append +00:00?
+        # Also, why did the format separate the date and time with `:`?
+        # END TODO
```
lol, these are all good questions. I imagine some of the contradictions probably came from John copying and pasting a docstring or changing the method name after already writing an [outdated] docstring. From experience, he wasn't always super careful with docstrings.

So, I think the best way to figure out what the intent of this was would be to run an actual load and add some logging to figure it out. I could look into doing that.

piehld reviewed (commented):
Thanks for bringing `mock-data` back @dmyersturnbull, and the other changes! I have a couple general questions about the removal of `TimeUtil.py`, which I left as comments for that file. I definitely agree with removing unnecessary dependencies from `requirements.txt`, but do wonder if we need to remove that file altogether or instead just update it, since that class is used in code elsewhere that would need to be updated also if we remove it. So, personally I'd kind of prefer keeping the class but just updating it to use `datetime` instead.
